### ISSUE TITLE
Inconsistent starting of service after crash/kill.

### ISSUE THREAD

**Issue by bilsch** [elastic/puppet-logstash#286]
https://github.com/elastic/puppet-logstash/blob/master/manifests/params.pp#L138

On rhel7 we are seeing the service enable change on every puppet run:

`/Stage[main]/Logstash::Service/Logstash::Service::Init[logstash]/Service[logstash]/enable: enable changed 'false' to 'true'`

the provider should be set to redhat not init ( then let puppet figure out which provider based on its semantics )

**Comment by jarpy** [comment]
Interesting, thanks.

I think there is something else happening here, and that the `$service_provider` variable is dead code that needs to be cleaned out. It's floating around the upper layers, but when we get down to the actual business of managing the service, it's disregarded.

https://github.com/elastic/puppet-logstash/blob/092c4b89c4a969f1a330a9e32be39bff76a68a17/manifests/service/init.pp#L127

That's mean there's now _two_ things I need to fix. :+1:

**Comment by alexizmailov** [comment]
Hi
I have the same issue AND "status => 'enabled | running" does not start the service.

>   class { 'logstash':
>     ensure       => 'present',
>     status       => 'enabled',
>     manage_repo  => true,
>     java_install => false,
>     autoupgrade  => false,
>   }
root@int-rsyslog-02:~ 130 # service logstash status
logstash is not running
root@int-rsyslog-02:~ 2 # puppet agent -t
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for int-rsyslog-02.somedomain.au
Info: Applying configuration version '2cd1ca8f0e1f76aafd432eaf7fee9f8eb43d1039'
Notice: /Stage[main]/Logstash::Service/Logstash::Service::Init[logstash]/Service[logstash]/enable: enable changed 'false' to 'true'
Notice: Applied catalog in 24.19 seconds
root@int-rsyslog-02:~  # service logstash status
logstash is not running

Centos 7, tested on logstash 2.1 and 2.3

**Comment by cdenneen** [comment]
@tylerlj any thoughts? This seems to have bitten me since upgrade to 2.3.3 or 0.11.0. Not sure which but service not running is very problematic. Woke up to a backlog of 67M that will take 36 hours to drain.

**Comment by jarpy** [comment]
There must be something pretty subtle going on here. After all, we have tests proving that the service is managed correctly on Centos 7.

https://travis-ci.org/elastic/puppet-logstash/jobs/140284810#L371

... so there must be something going on here which is not represented in the simple case.

**Comment by cdenneen** [comment]
If we make config change that notifies the service it comes up but if the
service dies it doesn't restart on next run. What's the current service
provider for EL7? systemd? I know you added this for Debian8.

**Comment by cdenneen** [comment]
Also provider of redhat only works for older EL <7. That's the init with
chkconfig. 7 and above need systemd.

**Comment by cdenneen** [comment]
Check this:
https://github.com/elastic/puppet-elasticsearch/blob/master/manifests/params.pp

**Comment by jarpy** [comment]
Are you referring to the `$service_providers` declared for each OS? That's dead code. It's not used for any logic that I can see.

https://github.com/elastic/puppet-logstash/search?utf8=%E2%9C%93&q=service_providers

In all cases except Debian-8 we let Puppet decide which provider to use. Maybe some Puppet versions are making the wrong choice.

Can you suggest a combination of Puppet version and OS version with which I can reproduce (and write tests) for this failure-mode? Thanks. :beers:

**Comment by cdenneen** [comment]
operatingsystemmajrelease >= 7 should be systemd.
Right now if you kill logstash process it will not be restarted by puppet.
If you make config change it will restart.

**Comment by jarpy** [comment]
Thanks. Which Puppet version? I have a hunch that the service provider is not being selected correctly/consistently.

**Comment by cdenneen** [comment]
4.3.2 (PE 2015.3)

**Comment by jarpy** [comment]
Awesome, I'm able to replicate this now, and here's something interesting...

Specifying `provider => systemd` does not resolve the problem.

```
[root@centos-7-x64 ~]# service logstash status
logstash is running

[root@centos-7-x64 ~]# pkill -f logstash

[root@centos-7-x64 ~]# service logstash status
logstash is not running

[root@centos-7-x64 ~]# puppet apply -e 'service { logstash: ensure => running }'
Notice: Compiled catalog for centos-7-x64.optusnet.com.au in environment production in 0.02 seconds
Notice: Applied catalog in 0.01 seconds

[root@centos-7-x64 ~]# puppet apply -e 'service { logstash: ensure => running, provider => systemd }'
Notice: Compiled catalog for centos-7-x64.optusnet.com.au in environment production in 0.02 seconds
Notice: Applied catalog in 0.01 seconds

[root@centos-7-x64 ~]# service logstash status
logstash is not running
```

Looking at the [source for the service provider](https://github.com/puppetlabs/puppet/blob/98d22be27447bcb34c33aa4a44610ab8768390f5/lib/puppet/provider/service/redhat.rb#L11), it seems that it's going to choose systemd for RedHat distros all the time. Those two datapoints suggest to me that the problem is not having the wrong provider, but something nasty elsewhere.

**Comment by jarpy** [comment]
This seems to be an interesting one, in which Logstash, Puppet and systemd all combine in weird ways to challenge us.

Logstash (version v2, this is [already fixed in v5-alpha3](https://github.com/elastic/logstash/issues/3606)) only provides a SysV init script, not unit files, so on systemd distributions we rely on SysV init backward-compatibility. That has some behaviour which is bad for us.

Look at the return code when we ask `systemctl` about a dead Logstash service:

```
root@localhost:~# killall -9 java

root@localhost:~# /etc/init.d/logstash status
logstash is not running

root@localhost:~# systemctl status logstash
● logstash.service - LSB: Starts Logstash as a daemon.
   Loaded: loaded (/etc/init.d/logstash)
   Active: active (exited) since Mon 2016-07-25 22:30:30 PDT; 10min ago
  Process: 10106 ExecStart=/etc/init.d/logstash start (code=exited, status=0/SUCCESS)

Jul 25 22:30:30 localhost logstash[10106]: logstash started.

root@localhost:~# echo $?
0
```

Well, I didn't expect to see exit code `0`, but then again, I didn't expect the service to be "active" either. I would have expected another adjective, like "hosed", "kaput" or "borken".

So, [how does Puppet decide](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/service/systemd.rb#L146) whether or not to start a dead service under systemd? It calls `systemctl is-active`, like this:

```
root@localhost:~# /etc/init.d/logstash status
logstash is not running

root@localhost:~# systemctl is-active logstash
active

root@localhost:~# echo $?
0
```

cc @untergeek, because he's been working very close to this problem, and probably understands what is going on with systemd (a thought that fills me with respect and awe).

**Comment by jarpy** [comment]
I should say, @untergeek, that I'm not asking for you to fix anything here (beyond what you already fixed in :sparkles:_the future_:sparkles:). I just thought you might be interested in this odd outcome.

**Comment by bilsch** [comment]
Sorry for staying so quiet on this guys. Have a look at https://tickets.puppetlabs.com/browse/PUP-2744 - may explain the core of the issue. Essentially its a bug in puppet

**Comment by jarpy** [comment]
No problem! I've been away on vacation myself, so didn't give this the attention it deserves.

Right now, my hope is that this will simply go away with the explicit support for systemd in Logstash 5. I'm going to test that hypothesis today. If it's true, I'll probably start working on an LS5 version of the module right away.

**Comment by jarpy** [comment]
Fixed by the work for #312.
